### PR TITLE
Full side chain import from dcrd on startup

### PR DESCRIPTION
**Pull request by chappjc**
This adds the ability to get all known side chains from dcrd and import all side chain blocks into the aux (dcrpg) database.

- [x] dcrd/rpcclient: add `GetChainTips` to RPC client (https://github.com/decred/dcrd/pull/1469)
- [x] rpcutils: add SideChainFull to get all blocks in a side chain
- [x] dcrpg: Modify StoreBlock and storeTxns to work correctly with side chain blocks.
- [x] dcrsqlite and dcrpg: Check for preexistence of side chain blocks in the DBs.
- [x] Capability to import all side chain blocks into dcrsqlite and dcrpg on start up.
- [x] Update/verify `/side` page for imported blocks, as currently it only shows side chain blocks from reorgs.

This resolves issue https://github.com/decred/dcrdata/issues/338.

For a follow-up PR:  Do the check and import periodically *while running* so that new forks that were not reorganized may be added without having to restart dcrdata.

### CODE REVIEW

**Comment by chappjc** [comment]
Tested on mainnet.  Successfully (apparently) imported 19 blocks from the 19 side chains known to the dcrd node used.  Blocks show up on the `/side` path and are all individually viewable.

Needs testing on testnet, and on other longer running nodes.

File: main.go [hunk 1]
```diff
@@ -400,14 +400,78 @@ func mainCore() error {
...
+				blockHash, err := chainhash.NewHashFromStr(hash)
+				if err != nil {
+					log.Errorf("Invalid block hash %s: %v.", hash, err)
+					continue
+				}
+
+				// Collect block data.
+				blockData, msgBlock, err := collector.CollectHash(blockHash)
+				if err != nil {
+					// Do not quite if unable to collect side chain block data.
```
Comment (dmigwi) [inline main.go]: quite -> quit

File: rpcutils/rpcclient.go [hunk 2]
```diff
@@ -261,6 +276,50 @@ func sideChainTips(allTips []dcrjson.GetChainTipsResult) (sideTips []dcrjson.Get
...
+
+	// Reverse side chain order so that last element is tip.
+	reverseStringSlice(sideChain)
+
+	return sideChain, nil
+}
+
+func reverseStringSlice(s []string) {
+	N := len(s)
+	for i := 0; i <= N/2-1; i++ {
```
Comment (dmigwi) [inline rpcutils/rpcclient.go]: I think its less confusing when you use bracket where you do multiple different operations;
 N/2-1 -> (N/2)-1

File: api/types/apitypes.go [hunk 3]
```diff
@@ -285,8 +285,15 @@ type BlockDataBasic struct {
...
 	Time       int64   `json:"time,omitemtpy"`
 	NumTx      uint32  `json:"txlength,omitempty"`
-	//TicketPoolInfo
-	PoolInfo TicketPoolInfo `json:"ticket_pool,omitempty"`
+	// TicketPoolInfo may be nil for side chain blocks.
+	PoolInfo *TicketPoolInfo `json:"ticket_pool,omitempty"`
+}
+
+// NewBlockDataBasic constructs a *BlockDataBasic with all fields allocated.
+func NewBlockDataBasic() *BlockDataBasic {
```
Comment (dmigwi) [inline api/types/apitypes.go]: "with all fields allocated" is confusing as you are just initializing `PoolInfo` field.

File: api/types/apitypes.go [hunk 4]
```diff
@@ -346,7 +353,15 @@ type StakeInfoExtended struct {
 	StakeDiff        float64              `json:"stakediff"`
 	PriceWindowNum   int                  `json:"window_number"`
 	IdxBlockInWindow int                  `json:"window_block_index"`
-	PoolInfo         TicketPoolInfo       `json:"ticket_pool"`
+	PoolInfo         *TicketPoolInfo      `json:"ticket_pool"`
+}
+
+// NewStakeInfoExtended constructs a *StakeInfoExtended with all fields
+// allocated.
```
Comment (dmigwi) [inline api/types/apitypes.go]: Same to this "all fields allocated" is confusing unless further explained.

File: blockdata/blockdata.go [hunk 5]
```diff
@@ -142,31 +141,39 @@ func (t *Collector) CollectAPITypes(hash *chainhash.Hash) (*apitypes.BlockDataBa
...
 	}
 	height := msgBlock.Header.Height
 	block := dcrutil.NewBlock(msgBlock)
 	txLen := len(block.Transactions())
+
+	// Coin supply and block subsidy
 	coinSupply, err := t.dcrdChainSvr.GetCoinSupply()
 	if err != nil {
 		log.Error("GetCoinSupply failed: ", err)
+		// Keep going to attemp acquisition of other block data.
```
[the remark this answers is not in the document]
Comment (dmigwi) [reply blockdata/blockdata.go]: This comment might also not be placed at the incorrect position.

File: db/dcrpg/pgblockchain.go [hunk 6]
```diff
@@ -180,6 +181,62 @@ func (db *ChainDBRPC) Store(blockData *blockdata.BlockData, msgBlock *wire.MsgBl
 	return db.ChainDB.Store(blockData, msgBlock)
 }
 
+// MissingSideChainBlocks identifies side chain blocks that are missing from the
+// DB. Side chains known to dcrd are listed via the getchaintips RPC. Each block
```
Comment (dmigwi) [inline db/dcrpg/pgblockchain.go]: Each block's presence is checked in the postgres DB

File: db/dcrpg/pgblockchain.go [hunk 7]
```diff
@@ -1826,10 +1885,15 @@ func (pgb *ChainDB) StoreBlock(msgBlock *wire.MsgBlock, winningTickets []string,
 	// Convert the wire.MsgBlock to a dbtypes.Block
 	dbBlock := dbtypes.MsgBlockToDBBlock(msgBlock, pgb.chainParams)
 
-	// Get the previous winners (stake DB pool info cache have this info)
+	// Get the previous winners (stake DB pool info cache have this info). If
+	// the previous block is side chain, stakedb will not have the
+	// winners/validators. Since Validators are only used to identify misses in
+	// InsertVotes, we will leave the Validators empty and assume there are no
+	// misses. TODO: ensure this wouldn't break reorg if this block becomes main
+	// chain at some point.
```
Comment (dmigwi) [inline db/dcrpg/pgblockchain.go]: If a block that is in a side chain can be brought back as a mainchain block what is the likelihood of this index being violated? 
```golang
IndexBlockTableOnHash = `CREATE UNIQUE INDEX uix_block_hash ON blocks(hash);`
```

File: db/dcrpg/pgblockchain.go [hunk 8]
```diff
@@ -1912,6 +1976,23 @@ func (pgb *ChainDB) StoreBlock(msgBlock *wire.MsgBlock, winningTickets []string,
 	lastBlockHash := msgBlock.Header.PrevBlock
 	// Only update if last was not genesis, which is not in the table (implied).
 	if lastBlockHash != zeroHash {
+		// Ensure previous block has the same main/sidechain status. If the
+		// curent block being added is side chain, do not invalidate the
```
Comment (dmigwi) [inline db/dcrpg/pgblockchain.go]: curent -> current

File: main.go [hunk 9]
```diff
@@ -334,30 +333,31 @@ func mainCore() error {
 
 	blockDataSavers = append(blockDataSavers, explore)
 
-	// Sync up with the blockchain
+	// Coordinated sync up both sqlite and postgres DBs with the network. This
```
Comment (dmigwi) [inline main.go]: Coordinated -> Coordinate. 
feels better.
"Coordinate the sync of both sqlite and postgres DBs with the network..."

If one day we decided to change the auxiliary db from Postgres to maybe Mysql or even a graph database, shall we be forced to rename the all Postgres occurrences to that specific db then. IMO I think we should use "auxiliary db" when referencing Postgres db now.

File: main.go [hunk 10]
```diff
@@ -334,30 +333,31 @@ func mainCore() error {
...
-		// available blocks to the baseDB. In return, baseDB maintains a
+		// Now that stakedb is either catching up or waiting for a block, start
+		// the auxDB sync, which is the master block getter, retrieving and
+		// making available blocks to the baseDB. In return, baseDB maintains a
 		// StakeDatabase at the best block's height.
 		go auxDB.SyncChainDBAsync(pgSyncRes, smartClient, quit,
 			updateAddys, updateVotes, newPGInds)
 
-		// Wait for the results
+		// Wait for the results from both DBs.
```
Comment (dmigwi) [inline main.go]: They are actually three database especially when full mode is running.

File: main.go [hunk 11]
```diff
@@ -400,14 +400,78 @@ func mainCore() error {
 		return nil
 	}
 
-	// Register for notifications from dcrd
+	// Ensure all side chains known by dcrd are also present in postgresql and
```
Comment (dmigwi) [inline main.go]: How about "auxiliary db" instead of "postgresql"?

File: db/dcrpg/pgblockchain.go [hunk 12]
```diff
@@ -180,6 +181,62 @@ func (db *ChainDBRPC) Store(blockData *blockdata.BlockData, msgBlock *wire.MsgBl
...
+		// For each block in the side chain, check if it already stored.
+		for is := range sideChain {
+			// Check for the block hash in the DB.
+			sideHeightDB, err := db.BlockHeight(sideChain[is])
+			if err != nil && err != sql.ErrNoRows {
+				// Unexpected error
+				log.Debugf("Failed to retrieve block %s.", sideChain[is])
+			} else if err == nil {
+				// This block is alredy in the DB.
+				log.Tracef("Found block %s in postgres at height %d.",
```
[the remark this answers is not in the document]
Comment (dmigwi) [reply db/dcrpg/pgblockchain.go]: I am talking about the comment

File: api/types/apitypes.go (hunk 3 above)
Comment (dmigwi) [reply api/types/apitypes.go]: "with valid pointer fields" feels better unless you've got a better option.

File: db/dcrpg/pgblockchain.go (hunk 7 above)
Comment (chappjc) [reply db/dcrpg/pgblockchain.go]: Being "brought back" does not involve the hash changing, so it would be violated if we tried to re-insert it, but we update the row with that hash so it won't be violated in a reorg.

The issue I was worried about is that there is no ticket pool info data ( for blocks imported this way (as side chain to start), and I was not certain that would be an issue if such a block somehow became mainchain via a reorg.

So, talking this through, the primary "pool info" not available to imported side chain blocks are the winning tickets (that vote in the next block) and the validators (the winning tickets that should be spent on votes in the block).  The validators are used in InsertVotes to identify missed votes as validators that did not vote.  Imported side chain blocks will just have no misses recorded in the misses table.  But if that block became main chain, the reorganization process will actually connect these blocks in stakedb, thus providing this information when `dcrpg.(*ChainMonitor).switchToSideChain` calls `(*ChainDB).StoreBlock` again, this time with `isMainchain=true`.  This time, `InsertVotes` will have the validators and will thus be able to insert any new misses.

To summarize, a reorganization will involve connecting the blocks in stakedb, allowing the validators to be known, and all the inserts will be run again, including the misses that were not known while the block was an imported side chain block.

File: main.go (hunk 10 above)
Comment (chappjc) [reply main.go]: Referring to the agenda db?  `getSyncd` only runs the sync routines of `wiredDB` and `ChainDB`.  Note that this keeps the `StakeDatabase` at the same height as these too, but that is managed by `wiredDB` so I'm only mentioning the two DBs with sync routines.
Do we need to be doing something with agenda db on startup, while catching up?

File: rpcutils/rpcclient.go (hunk 2 above)
Comment (chappjc) [reply rpcutils/rpcclient.go]: Personally I think this order of operations is pretty clear, but it's good to help readability if it doesn't affect code semantics.

File: main.go (hunk 10 above)
Comment (dmigwi) [reply main.go]: I would suggest we ignore agenda db for now as we have a lot of loose end that will be addressed when the new version of `getvoteinfo` rpc endpoint on dcrd is finalized.

Comment (dmigwi) [reply main.go]: But updates are always checked on startup https://github.com/decred/dcrdata/blob/master/main.go#L288-L294

File: db/dcrpg/pgblockchain.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (chappjc) [reply db/dcrpg/pgblockchain.go]: oh, lol :)

**Comment by chappjc** [comment]
Let me know if the last few commits address the issues you've identified. Thanks for the review.

**Comment by chappjc** [comment]
This resolves issue https://github.com/decred/dcrdata/pull/717.
